src/lattice.cpp

```diff
@@ -0,0 +1,945 @@
+      if (this_d < d) {
+        d = this_d;
+        if (w > 0) {
+          lattice_trans = {0, 0, 1};
+        } else {
+          lattice_trans = {0, 0, -1};
+        }
+      }
+    }
+  }
+
+  return {d, lattice_trans};
+}
+
+//==============================================================================
+
+std::array<int, 3>
+RectLattice::get_indices(const double xyz[3]) const
+{
+  int ix {static_cast<int>(std::ceil((xyz[0] - lower_left[0]) / pitch[0]))-1};
+  int iy {static_cast<int>(std::ceil((xyz[1] - lower_left[1]) / pitch[1]))-1};
+  int iz;
+  if (is_3d) {
+    iz = static_cast<int>(std::ceil((xyz[2] - lower_left[2]) / pitch[2]))-1;
+  } else {
+    iz = 0;
+  }
+  return {ix, iy, iz};
+}
+
+//==============================================================================
+
+std::array<double, 3>
+RectLattice::get_local_xyz(const double global_xyz[3], const int i_xyz[3]) const
+{
+  std::array<double, 3> local_xyz;
+  local_xyz[0] = global_xyz[0] - (lower_left[0] + (i_xyz[0] + 0.5)*pitch[0]);
+  local_xyz[1] = global_xyz[1] - (lower_left[1] + (i_xyz[1] + 0.5)*pitch[1]);
+  if (is_3d) {
+    local_xyz[2] = global_xyz[2] - (lower_left[2] + (i_xyz[2] + 0.5)*pitch[2]);
+  } else {
+    local_xyz[2] = global_xyz[2];
+  }
+  return local_xyz;
+}
+
+//==============================================================================
+
+int32_t&
+RectLattice::offset(int map, const int i_xyz[3])
+{
+  return offsets[nx*ny*nz*map + nx*ny*i_xyz[2] + nx*i_xyz[1] + i_xyz[0]];
+}
+
+//==============================================================================
+
+std::string
+RectLattice::index_to_string(int indx) const
+{
+  int iz {indx / (nx * ny)};
+  int iy {(indx - nx * ny * iz) / nx};
+  int ix {indx - nx * ny * iz - nx * iy};
+  std::string out {std::to_string(ix)};
+  out += ',';
+  out += std::to_string(iy);
+  if (is_3d) {
+    out += ',';
+    out += std::to_string(iz);
+  }
+  return out;
+}
+
+//==============================================================================
+
+void
+RectLattice::to_hdf5_inner(hid_t lat_group) const
+{
+  // Write basic lattice information.
+  write_string(lat_group, "type", "rectangular", false);
+  if (is_3d) {
+    write_double(lat_group, "pitch", pitch, false);
+    write_double(lat_group, "lower_left", lower_left, false);
+    write_int(lat_group, "dimension", n_cells, false);
+  } else {
+    std::array<double, 2> pitch_short {{pitch[0], pitch[1]}};
+    write_double(lat_group, "pitch", pitch_short, false);
+    std::array<double, 2> ll_short {{lower_left[0], lower_left[1]}};
+    write_double(lat_group, "lower_left", ll_short, false);
+    std::array<int, 2> nc_short {{n_cells[0], n_cells[1]}};
+    write_int(lat_group, "dimension", nc_short, false);
+  }
+
+  // Write the universe ids.  The convention here is to switch the ordering on
+  // the y-axis to match the way universes are input in a text file.
+  if (is_3d) {
+    hsize_t nx {static_cast<hsize_t>(n_cells[0])};
```

Comment:
I don't think a static_cast is necessary from `int` to `hsize_t`.

Reply:
It's a narrowing conversion so the compiler complains when debug flags are on.